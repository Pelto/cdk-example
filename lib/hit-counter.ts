import { RemovalPolicy, Duration } from "aws-cdk-lib";
import { Table, AttributeType, BillingMode } from "aws-cdk-lib/aws-dynamodb";
import { IFunction } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";

import * as path from 'path';

interface HitCounterProps {
  target: IFunction;
}

export class HitCounter extends Construct {

  public readonly proxyHandler: IFunction;

  public constructor(scope: Construct, id: string, props: HitCounterProps) {
    super(scope, id);

    const { target } = props;

    const table = new Table(this, "CounterTable", {
      partitionKey: { name: "Path", type: AttributeType.STRING },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    this.proxyHandler = new NodejsFunction(this, "HitCounterFunction", {
      entry: path.join(__dirname, "lambdas/hit-counter.ts"),
      timeout: Duration.seconds(15),
      environment: {
        tableName: table.tableName,
        downstreamFunction: target.functionName,
      },
      bundling: {
        minify: true,
        sourceMap: true,
      },
    });

    table.grantReadWriteData(this.proxyHandler);
    target.grantInvoke(this.proxyHandler);
  }
}
