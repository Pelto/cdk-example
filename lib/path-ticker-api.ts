import { RemovalPolicy, Duration } from "aws-cdk-lib";
import { LambdaRestApi, RestApi } from "aws-cdk-lib/aws-apigateway";
import { Table, AttributeType, BillingMode } from "aws-cdk-lib/aws-dynamodb";
import { IFunction } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";

import * as path from 'path';
interface HitCounterProps {
  handler: IFunction;
}
export class PathTickerApi extends Construct {

  public readonly api: RestApi;

  public constructor(scope: Construct, id: string, props: HitCounterProps) {
    super(scope, id);

    const { handler } = props;

    const table = new Table(this, 'CounterTable', {
      partitionKey: { name: 'Path', type: AttributeType.STRING },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const proxyHandler = new NodejsFunction(this, 'HitCounterFunction', {
      entry: path.join(__dirname, 'lambdas/hit-counter.ts'),
      timeout: Duration.seconds(15),
      environment: {
        tableName: table.tableName,
        downstreamFunction: handler.functionName,
      },
      bundling: {
        minify: true,
        sourceMap: true,
      },
    });

    table.grantReadWriteData(proxyHandler);
    handler.grantInvoke(proxyHandler);

    this.api = new LambdaRestApi(this, 'ApiGateway', {
      handler: proxyHandler,
    });
  }
}
