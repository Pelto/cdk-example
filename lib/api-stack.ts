import * as cdk from 'aws-cdk-lib';
import { CfnOutput, Duration, RemovalPolicy } from 'aws-cdk-lib';
import { LambdaRestApi, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { AttributeType, BillingMode, Table } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';

import * as path from "path";
import { HitCounter } from './hit-counter';

export class ApiStack extends cdk.Stack {

  public readonly apiEndpoint: CfnOutput;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const helloHandler = new NodejsFunction(this, "HelloFunction", {
      entry: path.join(__dirname, "lambdas/hello-world.ts"),
    });

    const hitCounter = new HitCounter(this, 'counter', {
      target: helloHandler,
    });

    const api = new LambdaRestApi(this, "Api", {
      handler: hitCounter.proxyHandler,
    });

    this.apiEndpoint = new CfnOutput(this, "ApiEndpoint", {
      value: api.url,
    });
  }
}
