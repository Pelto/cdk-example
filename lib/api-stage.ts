import { CfnOutput, Stage, StageProps } from "aws-cdk-lib";
import { Construct } from "constructs";

import { ApiStack } from "./api-stack";

export class ApiStage extends Stage {

  public readonly apiEndpoint: CfnOutput;

  public readonly hitTable: CfnOutput;

  public constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);

    const apiStack = new ApiStack(this, "ApiStack");

    this.apiEndpoint = apiStack.apiEndpoint;
    this.hitTable = apiStack.hitTable;
  }
}
