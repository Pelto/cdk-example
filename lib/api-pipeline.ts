import { RemovalPolicy, Stack, StackProps } from "aws-cdk-lib";
import { Repository } from "aws-cdk-lib/aws-codecommit";
import { Table } from "aws-cdk-lib/aws-dynamodb";
import { CodeBuildStep, CodePipeline, CodePipelineSource, ShellStep } from "aws-cdk-lib/pipelines";
import { Construct } from "constructs";

import { ApiStage } from "./api-stage";

export class ApiPipelineStack extends Stack{
  public constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const repository = new Repository(this, "Repository", {
      repositoryName: "api",
    });

    const pipeline = new CodePipeline(this, "Pipeline", {
      pipelineName: "ApiPipeline",
      synth: new CodeBuildStep("Synth", {
        input: CodePipelineSource.codeCommit(repository, "main"),
        installCommands: [
          `npm install -g aws-cdk`,
        ],
        buildEnvironment: {
          privileged: true,
        },
        commands: [
          `npm ci`,
          `npm run build`,
          `npx cdk synth`
        ]
      }),
    });

    const devApi = new ApiStage(this, "DevStage");
    const prodApi = new ApiStage(this, "ProdStage");

    const dev = pipeline.addStage(devApi);

    const testStep = new ShellStep("TestStep", {
      envFromCfnOutputs: {
        ENDPOINT_URL: devApi.apiEndpoint,
      },
      commands: [
        'curl --fail $ENDPOINT_URL/test',
      ],
    });

    dev.addPost(testStep);

    pipeline.addStage(prodApi);
  }
}
