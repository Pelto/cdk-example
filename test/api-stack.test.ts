import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ApiStack } from "../lib/api-stack";

test('ApiStack snapshot', () => {
  const app = new App();
  const pipeline = new ApiStack(app, 'Api', {});
  const template = Template.fromStack(pipeline);
  expect(template.toJSON()).toMatchSnapshot();
});
