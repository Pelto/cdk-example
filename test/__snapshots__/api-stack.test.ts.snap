// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiStack snapshot 1`] = `
Object {
  "Outputs": Object {
    "ApiApiGatewayEndpoint9EEF8DFF": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiApiGateway8FA53210",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiApiGatewayDeploymentStageprodCEA0AB12",
            },
            "/",
          ],
        ],
      },
    },
    "ApiEndpoint": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiApiGateway8FA53210",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiApiGatewayDeploymentStageprodCEA0AB12",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ApiApiGateway8FA53210": Object {
      "Properties": Object {
        "Name": "ApiGateway",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ApiApiGatewayANY54FC8F02": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "ApiHitCounterFunctionD29566DA",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "ApiApiGateway8FA53210",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "ApiApiGateway8FA53210",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiApiGatewayANYApiPermissionApiApiGateway3CDC40D9ANY7C2C0DA3": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ApiHitCounterFunctionD29566DA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiApiGateway8FA53210",
              },
              "/",
              Object {
                "Ref": "ApiApiGatewayDeploymentStageprodCEA0AB12",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiApiGatewayANYApiPermissionTestApiApiGateway3CDC40D9ANY00D4815F": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ApiHitCounterFunctionD29566DA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiApiGateway8FA53210",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiApiGatewayAccount903E1D79": Object {
      "DeletionPolicy": "Retain",
      "DependsOn": Array [
        "ApiApiGateway8FA53210",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "ApiApiGatewayCloudWatchRoleB185ED22",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiApiGatewayCloudWatchRoleB185ED22": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiApiGatewayDeploymentD1AFDABA2873eaafb691429fd685837897e3fec2": Object {
      "DependsOn": Array [
        "ApiApiGatewayproxyANYBC8B491D",
        "ApiApiGatewayproxy9A44C7DF",
        "ApiApiGatewayANY54FC8F02",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "ApiApiGateway8FA53210",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiApiGatewayDeploymentStageprodCEA0AB12": Object {
      "DependsOn": Array [
        "ApiApiGatewayAccount903E1D79",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "ApiApiGatewayDeploymentD1AFDABA2873eaafb691429fd685837897e3fec2",
        },
        "RestApiId": Object {
          "Ref": "ApiApiGateway8FA53210",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiApiGatewayproxy9A44C7DF": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "ApiApiGateway8FA53210",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": Object {
          "Ref": "ApiApiGateway8FA53210",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ApiApiGatewayproxyANYApiPermissionApiApiGateway3CDC40D9ANYproxy6D42C824": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ApiHitCounterFunctionD29566DA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiApiGateway8FA53210",
              },
              "/",
              Object {
                "Ref": "ApiApiGatewayDeploymentStageprodCEA0AB12",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiApiGatewayproxyANYApiPermissionTestApiApiGateway3CDC40D9ANYproxyF679F011": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ApiHitCounterFunctionD29566DA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiApiGateway8FA53210",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiApiGatewayproxyANYBC8B491D": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "ApiHitCounterFunctionD29566DA",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "ApiApiGatewayproxy9A44C7DF",
        },
        "RestApiId": Object {
          "Ref": "ApiApiGateway8FA53210",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiCounterTableB35878EF": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "Path",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "Path",
            "KeyType": "HASH",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "ApiHitCounterFunctionD29566DA": Object {
      "DependsOn": Array [
        "ApiHitCounterFunctionServiceRoleDefaultPolicy5B234BF6",
        "ApiHitCounterFunctionServiceRoleB2E3B293",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2f9a33e801715ddba5199a214a3c95583ac923609e6a45b0bf896e34f85ab512.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "downstreamFunction": Object {
              "Ref": "HelloFunctionD909AE8C",
            },
            "tableName": Object {
              "Ref": "ApiCounterTableB35878EF",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ApiHitCounterFunctionServiceRoleB2E3B293",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApiHitCounterFunctionServiceRoleB2E3B293": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiHitCounterFunctionServiceRoleDefaultPolicy5B234BF6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "ApiCounterTableB35878EF",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "HelloFunctionD909AE8C",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "HelloFunctionD909AE8C",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApiHitCounterFunctionServiceRoleDefaultPolicy5B234BF6",
        "Roles": Array [
          Object {
            "Ref": "ApiHitCounterFunctionServiceRoleB2E3B293",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HelloFunctionD909AE8C": Object {
      "DependsOn": Array [
        "HelloFunctionServiceRole186F1BA8",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b531685cd31e581aed03ace9733a728266d5ed8606f36364cf75c3afd6be7802.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "HelloFunctionServiceRole186F1BA8",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "HelloFunctionServiceRole186F1BA8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
